mlir/lib/Quantum/Transforms/wrapper_for_wrapper.cpp

```diff
@@ -0,0 +1,220 @@
+#include "mlir/Dialect/LLVMIR/LLVMDialect.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+#include "llvm/Support/FormatVariadic.h"
+
+#include "Quantum/Transforms/Passes.h"
+#include "Quantum/Transforms/Patterns.h"
+
+using namespace mlir;
+using namespace catalyst::quantum;
+
+namespace catalyst {
+namespace quantum {
+
+static size_t countCallsites(LLVM::LLVMFuncOp op)
+{
+    size_t count = 0;
+    op.walk([&](LLVM::CallOp callOp) { count++; });
+    return count;
+}
+
+static bool callsiteHasAttribute(LLVM::LLVMFuncOp op)
+{
+    bool retval = false;
+    ModuleOp moduleOp = cast<ModuleOp>(op->getParentOp());
+    if (!moduleOp)
+        return retval;
+    op.walk([&](LLVM::CallOp callOp) {
+        auto callee = callOp.getCallee();
+        if (callee) {
+            auto stringRef = callee.value().data();
+            Operation *fnDecl = moduleOp.lookupSymbol<LLVM::LLVMFuncOp>(stringRef);
+            retval = (bool)(fnDecl->getAttrOfType<UnitAttr>(
+                LLVM::LLVMDialect::getEmitCWrapperAttrName()));
+        }
+    });
+    return retval;
+}
+
+struct WrapperForWrapperTransform : public OpRewritePattern<LLVM::LLVMFuncOp> {
+    using OpRewritePattern<LLVM::LLVMFuncOp>::OpRewritePattern;
+
+    LogicalResult match(LLVM::LLVMFuncOp op) const override;
+    void rewrite(LLVM::LLVMFuncOp op, PatternRewriter &rewriter) const override;
+};
+
+LogicalResult WrapperForWrapperTransform::match(LLVM::LLVMFuncOp op) const
+{
+    std::string _mlir_ciface = "_mlir_ciface_";
+    size_t _mlir_ciface_len = _mlir_ciface.length();
+    auto symName = op.getSymName();
+    const char *symNameStr = symName.data();
+    size_t symNameLength = strlen(symNameStr);
+    // Filter based on name.
+    if (symNameLength <= _mlir_ciface_len)
+        return failure();
+
+    bool nameMatches = 0 == strncmp(symNameStr, _mlir_ciface.c_str(), _mlir_ciface_len);
+    if (!nameMatches)
+        return failure();
+
+    // Function must only contain a single function call.
+    size_t callsites = countCallsites(op);
+    if (callsites != 1)
+        return failure();
+
+    // The only call must contain the emitCWrapper attribute
+    bool hasAttribute = callsiteHasAttribute(op);
+
+    return hasAttribute ? success() : failure();
+}
+
+static bool functionHasReturns(LLVM::LLVMFuncOp op)
+{
+    bool result = false;
+    op.walk([&](LLVM::CallOp callOp) { result = (bool)callOp.getResult(); });
+    return result;
+}
+
+static bool functionHasInputs(LLVM::LLVMFuncOp op)
+{
+    bool result = false;
+    op.walk([&](LLVM::CallOp callOp) { result = !callOp.getOperandTypes().empty(); });
+    return result;
+}
+
+static LLVM::LLVMFunctionType
+convertFunctionTypeCatalystWrapper(PatternRewriter &rewriter, LLVM::LLVMFunctionType functionType,
+                                   bool hasReturns, bool hasInputs)
+{
+    SmallVector<Type, 2> transformedInputs;
+
+    Type ptr = LLVM::LLVMPointerType::get(rewriter.getContext());
```

Comment:
Small suggestion for better readability (or similar naming distinction between a value and pointer)
```suggestion
    Type ptr_t = LLVM::LLVMPointerType::get(rewriter.getContext());
```

Reply:
https://github.com/PennyLaneAI/catalyst/pull/106/commits/b0203d6da62011e4ac240f608af535591b553c2a